x/group/server/msg_server.go

```diff
@@ -0,0 +1,599 @@
+	"github.com/regen-network/regen-ledger/math"
+	"github.com/regen-network/regen-ledger/orm"
+	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/util"
+	"github.com/regen-network/regen-ledger/x/group"
+)
+
+func (s serverImpl) CreateGroup(ctx types.Context, req *group.MsgCreateGroupRequest) (*group.MsgCreateGroupResponse, error) {
+	metadata := req.Metadata
+	members := group.Members(req.Members)
+	admin := req.Admin
+
+	if err := members.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	maxMetadataLength := s.maxMetadataLength(ctx)
+	if err := assertMetadataLength(metadata, maxMetadataLength, "group metadata"); err != nil {
+		return nil, err
+	}
+
+	totalWeight := apd.New(0, 0)
+	for i := range members {
+		m := members[i]
+		if err := assertMetadataLength(m.Metadata, maxMetadataLength, "member metadata"); err != nil {
+			return nil, err
+		}
+
+		weight, err := math.ParseNonNegativeDecimal(m.Weight)
+		if err != nil {
+			return nil, err
+		}
+
+		if !weight.IsZero() {
+			err = math.Add(totalWeight, totalWeight, weight)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	groupID := group.ID(s.groupSeq.NextVal(ctx))
+	err := s.groupTable.Create(ctx, groupID.Bytes(), &group.GroupInfo{
+		GroupId:     groupID,
+		Admin:       admin,
+		Metadata:    metadata,
+		Version:     1,
+		TotalWeight: math.DecimalString(totalWeight),
+	})
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "could not create group")
+	}
+
+	for i := range members {
+		m := members[i]
+		err := s.groupMemberTable.Create(ctx, &group.GroupMember{
+			GroupId: groupID,
+			Member: &group.Member{
+				Address:  m.Address,
+				Weight:   m.Weight,
+				Metadata: m.Metadata,
+			},
+		})
+		if err != nil {
+			return nil, sdkerrors.Wrapf(err, "could not store member %d", i)
+		}
+	}
+
+	groupIDStr := util.Uint64ToBase58Check(groupID.Uint64())
+	err = ctx.EventManager().EmitTypedEvent(&group.EventCreateGroup{GroupId: groupIDStr})
+	if err != nil {
+		return nil, err
+	}
+
+	return &group.MsgCreateGroupResponse{GroupId: groupID}, nil
+}
+
+func (s serverImpl) UpdateGroupMembers(ctx types.Context, req *group.MsgUpdateGroupMembersRequest) (*group.MsgUpdateGroupMembersResponse, error) {
+	action := func(g *group.GroupInfo) error {
+		for i := range req.MemberUpdates {
+			groupMember := group.GroupMember{GroupId: req.GroupId,
+				Member: &group.Member{
+					Address:  req.MemberUpdates[i].Address,
+					Weight:   req.MemberUpdates[i].Weight,
+					Metadata: req.MemberUpdates[i].Metadata,
+				},
+			}
+			var found bool
+			var prevGroupMember group.GroupMember
+			switch err := s.groupMemberTable.GetOne(ctx, groupMember.NaturalKey(), &prevGroupMember); {
+			case err == nil:
+				found = true
+			case orm.ErrNotFound.Is(err):
+				found = false
+			default:
+				return sdkerrors.Wrap(err, "get group member")
+			}
+
+			totalWeight, err := math.ParseNonNegativeDecimal(g.TotalWeight)
+			if err != nil {
+				return err
+			}
```

Comment:
This should be outside of the loop. No need to parse and serialize every iteration